graph/src/runtime/wasm/module/context.rs

```diff
@@ -0,0 +1,1334 @@
+
+    // Used by ipfs.map.
+    pub(crate) timeout_stopwatch: Arc<std::sync::Mutex<TimeoutStopwatch>>,
+
+    // A trap ocurred due to a possible reorg detection.
+    pub possible_reorg: bool,
+
+    // A host export trap ocurred for a deterministic reason.
+    pub deterministic_host_trap: bool,
+
+    pub(crate) experimental_features: ExperimentalFeatures,
+
+    // This option is needed to break the cyclic dependency between, instance, store, and context.
+    // during execution it should always be populated.
+    asc_heap: Option<AscHeapCtx>,
+}
+
+impl WasmInstanceContextInner {
+    pub fn asc_heap_ref(&self) -> &AscHeapCtx {
+        self.asc_heap.as_ref().unwrap()
+    }
+    pub fn asc_heap_mut(&mut self) -> &mut AscHeapCtx {
+        self.asc_heap.as_mut().unwrap()
+    }
+}
+
+impl WasmInstanceContext {
+    pub fn set_asc_heap(&mut self, asc_heap: AscHeapCtx) {
+        self.inner.as_ref().borrow_mut().asc_heap = Some(asc_heap);
+    }
+
+    pub fn from_instance(
+        ctx: MappingContext,
+        valid_module: Arc<ValidModule>,
+        host_metrics: Arc<HostMetrics>,
+        timeout: Option<Duration>,
+        timeout_stopwatch: Arc<std::sync::Mutex<TimeoutStopwatch>>,
+        experimental_features: ExperimentalFeatures,
+    ) -> Result<Self, anyhow::Error> {
+        Ok(WasmInstanceContext {
+            inner: Rc::new(RefCell::new(WasmInstanceContextInner {
+                asc_heap: None,
+                ctx,
+                valid_module,
+                host_metrics,
+                timeout,
+                timeout_stopwatch,
+                possible_reorg: false,
+                deterministic_host_trap: false,
+                experimental_features,
+            })),
+        })
+    }
+
+    fn store_get_scoped(
+        &self,
+        store: &mut StoreContextMut<WasmInstanceContext>,
+        gas: &GasCounter,
+        entity_ptr: AscPtr<AscString>,
+        id_ptr: AscPtr<AscString>,
+        scope: GetScope,
+    ) -> Result<AscPtr<AscEntity>, HostExportError> {
+        let host_exports = self.as_ref().ctx.host_exports.cheap_clone();
+        let _timer = self
+            .as_ref()
+            .host_metrics
+            .cheap_clone()
+            .time_host_fn_execution_region("store_get");
+
+        let read_store = store.as_context();
+        let entity_type: String =
+            asc_get(&read_store, self.as_mut().asc_heap_mut(), entity_ptr, gas)?;
+        let id: String = asc_get(&read_store, self.as_mut().asc_heap_mut(), id_ptr, gas)?;
+        let entity_option = host_exports
+            .store_get(
+                &mut self.as_mut().ctx.state,
+                entity_type.clone(),
+                id.clone(),
+                gas,
+                scope,
+            )?
+            // This is not great in a hot path but otherwise the self ref would not
+            // be released for the next block. Would be good to find a better pattern here.
+            .map(|e| e.into_owned());
```

Comment:
This change builds for me without requiring a clone:

```suggestion
        let block_state = &mut self.as_mut().ctx.state;
        let entity_option =
            host_exports.store_get(block_state, entity_type.clone(), id.clone(), gas, scope)?;
```

Reply:
it compiles but breaks the tests because of the runtime borrow checking so I had to put it back

Reply:
I found a way around the clone, but it's ugly as hell. The clone basically undoes PR #4624. Do we really need interior mutability for `WasmInstanceContext`? Can't we get away with just plain `&mut` as we used to?

```diff
diff --git a/graph/src/runtime/wasm/module/context.rs b/graph/src/runtime/wasm/module/context.rs
index cc3733214..6502209c3 100644
--- a/graph/src/runtime/wasm/module/context.rs
+++ b/graph/src/runtime/wasm/module/context.rs
@@ -140,6 +140,11 @@ impl WasmInstanceContext {
         scope: GetScope,
     ) -> Result<AscPtr<AscEntity>, HostExportError> {
         let host_exports = self.as_ref().ctx.host_exports.cheap_clone();
+        let instrument = self.as_ref().ctx.instrument;
+        let host_metrics = self.as_ref().host_metrics.cheap_clone();
+        let debug_fork = self.as_ref().ctx.debug_fork.cheap_clone();
+        let logger = self.as_ref().ctx.logger.cheap_clone();
+
         let _timer = self
             .as_ref()
             .host_metrics
@@ -150,37 +155,29 @@ impl WasmInstanceContext {
         let entity_type: String =
             asc_get(&read_store, self.as_mut().asc_heap_mut(), entity_ptr, gas)?;
         let id: String = asc_get(&read_store, self.as_mut().asc_heap_mut(), id_ptr, gas)?;
+        let (mut heap_mut, mut state_mut) = RefMut::map_split(
+            self.as_mut(),
+            |WasmInstanceContextInner { ctx, asc_heap, .. }| {
+                (asc_heap.as_mut().unwrap(), &mut ctx.state)
+            },
+        );
         let entity_option = host_exports
-            .store_get(
-                &mut self.as_mut().ctx.state,
-                entity_type.clone(),
-                id.clone(),
-                gas,
-                scope,
-            )?
+            .store_get(&mut state_mut, entity_type.clone(), id.clone(), gas, scope)?
             // This is not great in a hot path but otherwise the self ref would not
             // be released for the next block. Would be good to find a better pattern here.
             .map(|e| e.into_owned());
 
-        if self.as_ref().ctx.instrument {
-            debug!(self.as_ref().ctx.logger, "store_get";
+        if instrument {
+            debug!(logger, "store_get";
                     "type" => &entity_type,
                     "id" => &id,
                     "found" => entity_option.is_some());
         }
-        let host_metrics = self.as_ref().host_metrics.cheap_clone();
-        let debug_fork = self.as_ref().ctx.debug_fork.cheap_clone();
 
         let ret = match entity_option {
             Some(entity) => {
                 let _section = host_metrics.stopwatch.start_section("store_get_asc_new");
-                // let entity = entity.to_owned().sorted_ref();
-                asc_new(
-                    store,
-                    self.as_mut().asc_heap_mut(),
-                    &entity.sorted_ref(),
-                    gas,
-                )?
+                asc_new(store, &mut *heap_mut, &entity.sorted_ref(), gas)?
             }
             None => match &debug_fork {
                 Some(fork) => {
@@ -194,12 +191,7 @@ impl WasmInstanceContext {
                         Some(entity) => {
                             let _section =
                                 host_metrics.stopwatch.start_section("store_get_asc_new");
-                            let entity = asc_new(
-                                store,
-                                self.as_mut().asc_heap_mut(),
-                                &entity.sorted(),
-                                gas,
-                            )?;
+                            let entity = asc_new(store, &mut *heap_mut, &entity.sorted(), gas)?;
                             self.store_set(store, gas, entity_ptr, id_ptr, entity)?;
                             entity
                         }
```

Reply:
Even removing the RefCell this borrow conflict persisted (but now at compile time). I fixed it by switching from `Cow` to `Arc`.